### PR TITLE
fix a bug in using record_host_persistent_memory_allocation

### PR DESCRIPTION
**Pull request by snnn**

### CODE REVIEW

**Comment by snnn** [comment]
Indeed, this bug only affects 32 bits builds. Is there any one using 32-bits tensorflow?

**Comment by yaroslavvb** [comment]
@samjabrahams are Raspberry Pi TF builds 32 bit by any chance?

**Comment by samjabrahams** [comment]
@yaroslavvb yes- though unfortunately I won't be able to play around with the effect of this code for a week or two (traveling).

Is this the gist of the problem: casting between `size_t` and `int64` on 32-bit systems causes memory to not be freed/tracked properly? Just trying to make sure I'm following.

**Comment by snnn** [comment]
The following code:

	#include <iostream>
	#include <stdint.h>

	void record_host_persistent_memory_allocation(int64_t size) {
		std::cout << size << std::endl;
	}
	int main() {
		size_t a = 3;
		record_host_persistent_memory_allocation(-a);
		return 0;
	}

output is "4294967293" on 32 bits build and "-3" on 64 bits build.

**Comment by samjabrahams** [comment]
I'll take a looksy when I am near some hardware. I also need to double check exactly which flags are set when TensorFlow compiles.

File: tensorflow/core/kernels/reduction_ops_common.h [hunk 1]
```diff
@@ -238,9 +238,9 @@ class ReductionOp : public OpKernel {
     if (ctx->track_allocations()) {
       // The temporary memory becomes the output memory.
       if (ctx->allocate_on_host(alloc_attr)) {
-        ctx->record_host_temp_memory_size(-out.AllocatedBytes());
+        ctx->record_host_temp_memory_size(-(int64)out.AllocatedBytes());
```
Comment (rmlarsen) [inline tensorflow/core/kernels/reduction_ops_common.h]: Please use C++ style casts, i.e. static_cast<int64>(foo).

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/reduction_ops_common.h (hunk 1 above)
Comment (snnn) [inline tensorflow/core/kernels/reduction_ops_common.h]: @rmlarsen fixed.

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Review by rmlarsen — APPROVED** [review]

**Pull request merged**